src/settings/SettingsSelfExclusion.js

```diff
@@ -48,6 +36,28 @@ export default class SettingsSelfExclusion extends PureComponent {
     this.validationMan = new ValidationManager(this.constraints);
   }
 
+  getFormData(props) {
+    return {
+      max_balance: props.max_balance,
+      max_turnover: props.max_turnover,
+      max_losses: props.max_losses,
+      max_7day_turnover: props.max_7day_turnover,
+      max_7day_losses: props.max_7day_losses,
+      max_30day_turnover: props.max_30day_turnover,
+      max_30day_losses: props.max_30day_losses,
+      max_open_bets: props.max_open_bets,
+      session_duration_limit: props.session_duration_limit,
+      timeout_until_time: props.timeout_until,
+      timeout_until_date: props.timeout_until,
+      exclude_until: props.exclude_until,
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
```

Comment:
@brucebinary I think componentWillRecieveProps is fired even when props are not changed so when using that for preventing extra rerendering we need to check if current props are changed, what do you think about checking (next props !== this.props) before doing anything?

Reply:
@nazaninreihani "I think componentWillRecieveProps is fired even when props are not changed" are you sure about that? This is to challenge Facebook's architecture.

Reply:
@arnabk it's on their own document 
https://reactjs.org/docs/react-component.html#componentwillreceiveprops

Reply:
it's odd to need that condition. By adding that condition we are comparing the addresses of those 2 objects, which will always be different (I've check this just to be sure). Unless what you mean is to compare each individual attribute of the form data (like how it was initially implemented by @arnabk ):

```
  componentWillReceiveProps = (newProps) => {
    const newState = {};
    if (newProps.max_balance !== this.state.max_balance) newState.max_balance = newProps.max_balance;
    if (newProps.max_turnover !== this.state.max_turnover) newState.max_turnover = newProps.max_turnover;
    if (newProps.max_losses !== this.state.max_losses) newState.max_losses = newProps.max_losses;
    if (newProps.max_7day_turnover !== this.state.max_7day_turnover) newState.max_7day_turnover = newProps.max_7day_turnover;
    if (newProps.max_7day_losses !== this.state.max_7day_losses) newState.max_7day_losses = newProps.max_7day_losses;
    if (newProps.max_30day_turnover !== this.state.max_30day_turnover) newState.max_30day_turnover = newProps.max_30day_turnover;
    if (newProps.max_open_bets !== this.state.max_open_bets) newState.max_open_bets = newProps.max_open_bets;
    if (newProps.session_duration_limit !== this.state.session_duration_limit) newState.session_duration_limit = newProps.session_duration_limit;
    if (newProps.timeout_until_time !== this.state.timeout_until_time) newState.timeout_until_time = newProps.timeout_until_time;
    if (newProps.timeout_until_date !== this.state.timeout_until_date) newState.timeout_until_date = newProps.timeout_until_date;
    if (newProps.exclude_until !== this.state.exclude_until) newState.exclude_until = newProps.exclude_until;
    if (Object.keys(newState).length > 0) this.setState(newState);
  }

``` 

that would be pretty verbose, and I'm not too concern with rerenders in virtual DOM as it is pretty cheap to do. Rerenders also do not happen immediately; react may delay them and update a few changes in a single pass.

However I am willing to place the condition block back if everyone agrees this is more performant. @aminroosta your thoughts?

Reply:
ops. I didn't know there were comments before mine. Github doesn't update this page in the background

Reply:
@brucebinary @arnabk what happens to state if componentWillReceiveProps is called unexpectedly in the middle of editing the field by the user? isn't the state going to be set to new Props and all user input missing?

Reply:
@nazaninreihani Isn't PureComponent supposed to skip the rerendering if the props are same? I thought that is the reason they introduced it instead of we doing the shallow compare.

Reply:
@nazaninreihani I've recently given some thought to that as well. I've set a breakpoint on `componentWillReceiveProps` before, and noticed that the form data attributes were `undefined` for 3 consecutive times, and then contain data in the next 3 consecutive calls - this happens so fast that the user never notices, but it could be probable that in slow connections user can have his inputs overwritten (not with blank data but saved data).

I find this behaviour very unusual. Shouldn't the form data in props be there during initialisation and not only receive the data in the background milliseconds later? Filling the form data using `componentWillReceiveProps` is a little odd to me.

Reply:
@arnabk @brucebinary Found the issue, the issue seems when we set initial state from props, after recieving props the new props are not passed to the component, so we can remove the initial value for formData from constructor and set inital values in render

Reply:
@brucebinary sorry I didn't see your last commits